quickwit-search/src/client_pool/search_client_pool.rs

```diff
@@ -0,0 +1,170 @@
+        let client_pool = SearchClientPool {
+            clients: Arc::new(RwLock::new(clients)),
+        };
+
+        // Prepare to start a thread that will monitor cluster members.
+        let thread_clients = Arc::clone(&client_pool.clients);
+        let mut members_watch_channel = cluster.member_change_watcher();
+
+        // Start to monitor the cluster members.
+        tokio::spawn(async move {
+            while let Some(members) = members_watch_channel.next().await {
+                let mut clients = thread_clients.write().await;
+
+                // Create a list of addresses to be removed.
+                let members_addresses: HashSet<SocketAddr> = members
+                    .iter()
+                    .map(|member| get_grpc_addr(member.listen_addr))
+                    .collect();
+                let addrs_to_remove: Vec<SocketAddr> = clients
+                    .keys()
+                    .filter(|socket_addr| !members_addresses.contains(*socket_addr))
+                    .cloned()
+                    .collect();
+
+                // Remove clients from the client pool.
+                for grpc_addr in addrs_to_remove {
+                    let removed = clients.remove(&grpc_addr).is_some();
+                    if removed {
+                        debug!(grpc_addr=?grpc_addr, "Remove a client that is connecting to the node that has been downed or left the cluster.");
+                    }
+                }
+
+                // Add clients to the client pool.
+                for member in members {
+                    let grpc_addr = get_grpc_addr(member.listen_addr);
+                    if let Entry::Vacant(_entry) = clients.entry(grpc_addr) {
+                        match create_search_service_client(grpc_addr).await {
+                            Ok(client) => {
+                                debug!(grpc_addr=?grpc_addr, "Add a new client that is connecting to the node that has been joined the cluster.");
+                                clients.insert(grpc_addr, client);
+                            }
+                            Err(err) => {
+                                error!(grpc_addr=?grpc_addr, err=?err, "Failed to create search client.")
+                            }
+                        };
+                    }
+                }
+            }
+        });
+
+        Ok(client_pool)
+    }
+}
+
+#[async_trait]
+impl ClientPool for SearchClientPool {
+    /// Assign the given job to the clients.
+    /// Returns a list of pair (SocketAddr, Vec<Job>)
+    async fn assign_jobs(&self, mut jobs: Vec<Job>) -> anyhow::Result<Vec<(SocketAddr, Vec<Job>)>> {
+        let mut splits_groups: HashMap<SocketAddr, Vec<Job>> = HashMap::new();
+
+        // Distribute using rendez-vous hashing
+        let clients = self.clients.read().await;
+        let mut nodes: Vec<Node> = clients
+            .iter()
+            .map(|(grpc_addr, _client)| Node::new(&grpc_addr.to_string(), 0))
```

Comment:
Can we keep the SocketAddr in the Node and avoid the serialization/deserialization?